File: ParkingRegisterAPI/ParkingRegistry/SlotRegistry.cs
using ParkingRegisterAPI.Models;

namespace ParkingRegisterAPI.ParkingRegistry
{
    public class SlotRegistry
    {
        public static void SetupLots(int numberOfLots)
        {
            for (int i = 1; i <= numberOfLots; i++)
            {
                Slot lot = new Slot
                {
                    Number = i.ToString()
                };
                __slots[lot.Number] = lot;
            }
        }

        private static Dictionary<string, Slot> __slots { get; set; } = new Dictionary<string, Slot>();

        public static string FindEmptySlotNumber()
        {
            return __slots.FirstOrDefault(lotKvp => lotKvp.Value.IsEmpty()).Key;
        }

        public static void ParkCar(string slotNumber, Car car)
        {
            __slots[slotNumber].ParkCar(car);
        }

        public static void UnparkCar(string slotNumber)
        {
            __slots[slotNumber].UnparkCar();
        }

        public static Slot GetSlot(string slotNumber)
        {
            if (!__slots.ContainsKey(slotNumber)) return null;
            return __slots[slotNumber];
        }
    }
}

File: ParkingRegisterAPIFS/ParkingRegistry/SlotRegistry.fs
namespace ParkingRegisterAPIFS.ParkingRegistry

open ParkingRegisterAPIFS.Models.CarInSlot
open ParkingRegisterAPIFS.Models.Slot
open ParkingRegisterAPIFS.Models.Car
open System

module SlotRegistry =
    type ISlotRegistry =
        abstract member findEmptySlotNumber: unit -> string
        abstract member parkCar: slotNumber: string * Car -> unit
        abstract member unparkCar: string -> unit
        abstract member getSlot: string -> Nullable<Slot>


    type LotSizeInfo() =
        static let mutable __numberOfLots: int = 0
        
        static member setupLots(numberOfLots: int) =
            __numberOfLots <- numberOfLots

        static member getNumberOfLots() =
            __numberOfLots


    type SlotRegistry() =
        inherit LotSizeInfo()

        let mutable __slots = Lazy<Map<string,Slot>>.Create(
            fun() -> seq {for slotNumber in 1 .. LotSizeInfo.getNumberOfLots() do yield new Slot(string slotNumber)} |>
                        Seq.map(fun slot -> slot.Number, slot) |>
                        Map.ofSeq)

        let changeSlot(slotNumber: string, carInSlot: Nullable<CarInSlot>) =
            Lazy<Map<string,Slot>>(__slots.Value |> Map.change slotNumber (
                fun slot -> match slot with
                            | Some value -> Some (Slot(value.Number, carInSlot))
                            | None -> None))


        interface ISlotRegistry with
            member _.findEmptySlotNumber() =
                __slots.Value |> Map.tryFindKey (fun _ value -> not value.CarInSlot.HasValue) |> Option.defaultValue ""

            member _.parkCar(slotNumber: string, car: Car) =
                __slots <- changeSlot(slotNumber, car.CarInSlot)

            member _.unparkCar(slotNumber: string) =
                __slots <- changeSlot(slotNumber, Nullable<CarInSlot>())

            member _.getSlot(slotNumber: string) =
                if __slots.Value |> Map.containsKey slotNumber then
                    Nullable<Slot>(__slots.Value.[slotNumber])
                else
                    Nullable<Slot>()
